ssh/scp_test.go

```diff
@@ -82,54 +83,114 @@ func TestCopy(t *testing.T) {
-//		source     string
-//		shouldFail bool
-//	}{
-//		{
-//			name:   "user and host",
-//			args:   SSHArgs{User: "sshuser", Host: "local.host"},
-//			source: "/tmp/any",
-//			cmdStr: "scp -rpq -o StrictHostKeyChecking=no -P 22 sshuser@local.host:/tmp/any",
-//		},
-//		{
-//			name:   "user host and pkpath",
-//			args:   SSHArgs{User: "sshuser", Host: "local.host", PrivateKeyPath: "/pk/path"},
-//			source: "/foo/bar",
-//			cmdStr: "scp -rpq -o StrictHostKeyChecking=no -i /pk/path -P 22 sshuser@local.host:/foo/bar",
-//		},
-//		{
-//			name:   "user host pkpath and proxy",
-//			args:   SSHArgs{User: "sshuser", Host: "local.host", PrivateKeyPath: "/pk/path", ProxyJump: &ProxyJumpArgs{User: "juser", Host: "jhost"}},
-//			source: "userFile",
-//			cmdStr: "scp -rpq -o StrictHostKeyChecking=no -i /pk/path -P 22 -J juser@jhost sshuser@local.host:userFile",
-//		},
-//		{
-//			name:       "missing host",
-//			args:       SSHArgs{User: "sshuser"},
-//			shouldFail: true,
-//		},
-//	}
-//
-//	for _, test := range tests {
-//		t.Run(test.name, func(t *testing.T) {
-//			result, err := makeSCPCmdStr("scp", test.args, test.source)
-//			if err != nil && !test.shouldFail {
-//				t.Fatal(err)
-//			}
-//			cmdFields := strings.Fields(test.cmdStr)
-//			resultFields := strings.Fields(result)
-//
-//			for i := range cmdFields {
-//				if cmdFields[i] != resultFields[i] {
-//					t.Fatalf("unexpected command string element: %s vs. %s", cmdFields, resultFields)
-//				}
-//			}
-//		})
-//	}
-//}
+func TestCopyTo(t *testing.T) {
+	tests := []struct {
+		name        string
+		sshArgs     SSHArgs
+		localFiles  map[string]string
+		file        string
+		fileContent string
+	}{
+		{
+			name:        "copy single file to remote",
+			sshArgs:     testSSHArgs,
+			localFiles:  map[string]string{"local-foo.txt": "FooBar"},
+			file:        "local-foo.txt",
+			fileContent: "FooBar",
+		},
+		{
+			name:        "copy single file in dir to remote",
+			sshArgs:     testSSHArgs,
+			localFiles:  map[string]string{"local-foo/local-bar.txt": "FooBar"},
+			file:        "local-foo/local-bar.txt",
+			fileContent: "FooBar",
+		},
+		{
+			name:        "copy dir entire dir to remote",
+			sshArgs:     testSSHArgs,
+			localFiles:  map[string]string{"local-bar/local-foo.csv": "FooBar", "local-bar/local-bar.txt": "BarBar"},
+			file:        "local-bar/",
+			fileContent: "FooBar",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			defer func() {
+				for file := range test.localFiles {
+					RemoveLocalTestFile(t, filepath.Join(support.TmpDirRoot(), file))
+					RemoveRemoteTestSSHFile(t, test.sshArgs, file)
+				}
+			}()
+
+			// setup fake local files
+			for file, content := range test.localFiles {
+				MakeLocalTestFile(t, filepath.Join(support.TmpDirRoot(), file), content)
+			}
+
+			// create remote dir if needed
+			// setup remote dir if needed
+			MakeRemoteTestSSHDir(t, test.sshArgs, test.file)
```

Comment:
Does `copy_to` fail if the target directory does not exits?
Since it is just scp I would guess it does not. 

Do we need this? if we do, then we should probably add a test for the other case as well.

Reply:
No, right now, it relies on the target to be there.  However, if target is not provided, it sets target to source path.